compiler/sem/semstmts.nim

```diff
@@ -462,11 +462,508 @@ proc errorSymChoiceUseQualifier(c: PContext; n: PNode) =
+        # TODO: convert to nkError
+        defaultConstructionError(c, v.typ, v.info)
+      else:
+        # TODO: convert to nkError
+        checkNilable(c, v)
+      
+      # allow let to not be initialised if imported from C:
+      if v.kind == skLet and sfImportc notin v.flags:
+        # TODO: make nkError
+        localReport(c.config, defPart, reportSem rsemLetNeedsInit)
+    
+    if sfCompileTime in v.flags:
+      var x = newNodeI(result.kind, v.info)
+      x.add producedDecl
+      vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
+
+    if v.flags * {sfGlobal, sfThread} == {sfGlobal}:
+      # this is just logging, doesn't need to be converted to an error
+      localReport(c.config, v.info, reportSym(rsemGlobalVar, v))
+
+  case defPart.kind
+  of nkVarTuple:
+    # these transform `b` into an `nkError` if required, we use this when
+    # populating all the children
+    result.add:
+      if tupTyp.kind != tyTuple:
+        hasError = true
+        newError(
+              c.config,
+              producedDecl,
+              c.config.semReportTypeMismatch(producedDecl, {tyTuple}, tupTyp))
+      elif sameTupleTypeAndDefAirity:
+        hasError = true
+        newError(
+              c.config,
+              producedDecl,
+              semReportCountMismatch(
+                rsemWrongNumberOfVariables,
+                expected = defCount,
+                got = tupTyp.len, node = producedDecl))
+      else:
+        producedDecl
+
+  of nkIdentDefs:
+    result.add producedDecl
+
+    if tupTyp.kind == tyTuple and def.kind in {nkPar, nkTupleConstr} and
+        defPart.len > 3:
+      # xxx: helpful hints like this should likely be associated to an id/pos
+      #      instead of reporting
+      localReport(c.config, defPart.info, reportSem rsemEachIdentIsTuple)
+  else:
+    discard
+
+  if hasError:
+    # xxx: hasError is here until we've converted the whole proc to use nkError
+    #      until then we check at the end to ensure we capture any nkErrors
+    #      embedded within the section.
+    #
+    #      we also need to use the conditional version `wrapIfErrorInSubTree`
+    #      because we use `isErrorLike` earlier in the proc and that means we
+    #      don't necessarily have an nkError node, but it could be an error
+    #      symbol or type.
+    result = c.config.wrapIfErrorInSubTree(result)
+
+
+proc semLetOrVar(c: PContext, n: PNode, symkind: TSymKind): PNode =
+  ## semantically analyses let or var sections, analysis follows these steps:
+  ## 0. results are accumulated into a statement list and holds the section
+  ## 1. for each entry in the section to analyse:
+  ##    - if a single ident in the identdef it's analysed
+  ##    - multiple idents are decomposed and analysed one at a time
+  ##    - nkVarTuple:
+  ##      1. rhs assigns to a single intermediary identdef and then analysed
+  ##      2. for each unpack item, one identdef, analysed one at a time
```

Comment:
I'm going to drop the lowering approach. Pretty sure I figured out a way around it.